source/host_resolver.c

```diff
@@ -547,6 +567,61 @@ static inline void process_records(
     }
 }
 
+static void s_purge_host_cache_callback_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct host_purge_callback_options *options = arg;
+    aws_mem_release(options->allocator, task);
+    options->on_host_purge_complete(options->on_host_purge_complete_user_data);
+    aws_mem_release(options->allocator, options);
+}
+
+static int resolver_purge_host_cache(const struct aws_host_resolver_purge_host_options *options) {
+    struct default_host_resolver *default_host_resolver = options->resolver->impl;
+
+    AWS_LOGF_INFO(AWS_LS_IO_DNS, "id=%p: purging record for %s", (void *)options->resolver, options->host->bytes);
+
+    aws_mutex_lock(&default_host_resolver->resolver_lock);
+
+    struct aws_hash_element *element = NULL;
+    aws_hash_table_find(&default_host_resolver->host_entry_table, options->host, &element);
+
+    /* Success if entry doesn't exist in cache. */
+    if (element == NULL) {
+        aws_mutex_unlock(&default_host_resolver->resolver_lock);
+        if (options->on_host_purge_complete_callback != NULL) {
+            /* Schedule completion callback asynchronouly */
+            struct host_purge_callback_options *purge_callback_options =
+                aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct host_purge_callback_options));
+            purge_callback_options->allocator = default_host_resolver->allocator;
+            purge_callback_options->on_host_purge_complete = options->on_host_purge_complete_callback;
+            purge_callback_options->on_host_purge_complete_user_data = options->on_host_purge_complete_user_data;
+
+            struct aws_task *task = aws_mem_calloc(default_host_resolver->allocator, 1, sizeof(struct aws_task));
+            aws_task_init(
+                task, s_purge_host_cache_callback_task, purge_callback_options, "async_purge_host_address_task");
+
+            struct aws_event_loop *loop = aws_event_loop_group_get_next_loop(default_host_resolver->event_loop_group);
+            aws_event_loop_schedule_task_now(loop, task);
+        }
+        return AWS_OP_SUCCESS;
+    }
+
+    struct host_entry *host_entry = element->value;
+    AWS_FATAL_ASSERT(host_entry);
+
+    /* Setup the on_host_purge_complete callback. */
+    aws_mutex_lock(&host_entry->entry_lock);
+    host_entry->on_host_purge_complete = options->on_host_purge_complete_callback;
+    host_entry->on_host_purge_complete_user_data = options->on_host_purge_complete_user_data;
```

Comment:
Is there any chance for `host_entry->on_host_purge_complete` already be set? If not, can we ASSERT on that? I think that will be more clear in the code that the callback will not be override and lead to a never-invoked complete callback.

Reply:
I have added an assert. We remove it from the hash table after setting these. So next time we will not find the entry in the cache.